File: sdks/packages/typescript/examples/01-list-devices.ts
/**
 * Example 01 — List devices.
 *
 * The most basic thing you can do: construct a client and ask the daemon which
 * devices it can see. This talks to `GET /list` under the hood via
 * `client.devices.list()` and does NOT require a device to be attached (an empty
 * list is a perfectly valid, successful response).
 *
 * Run it on its own:
 *   GO_IOS_API_KEY=... npx tsx examples/01-list-devices.ts
 */
import { isMain, makeClient, runAsScript, type IosClient } from "./_shared";

/**
 * The suite runner imports and calls this `run(client)` so all examples share
 * one client and the results can be aggregated. Each example follows the same
 * shape: an exported `run(client)` plus a direct-execution guard at the bottom.
 */
export async function run(client: IosClient): Promise<void> {
  // `list()` returns a { deviceList } envelope. Each entry carries the device's
  // udid at `properties.serialNumber` — that's the id every device-scoped route
  // keys on.
  const { deviceList } = await client.devices.list();

  console.log(`Found ${deviceList.length} device(s):`);
  for (const entry of deviceList) {
    const udid = entry.properties.serialNumber;
    // `connectionType` (e.g. "USB", "Network") and `address` (for tunneled
    // devices) are optional metadata the daemon includes when known.
    const via = entry.properties.connectionType ?? "unknown";
    const net = entry.address ? ` @ ${entry.address}` : "";
    console.log(`  - ${udid} (via ${via})${net}`);
  }

  if (deviceList.length === 0) {
    // Not an error: the daemon is up and answered, there's just nothing plugged
    // in. The suite runner treats this example as a PASS regardless.
    console.log("(no devices attached — plug one in to see it here)");
  }
}

// Direct execution: `tsx examples/01-list-devices.ts`.
if (isMain(import.meta.url)) {
  await runAsScript("01-list-devices", () => run(makeClient()));
}

File: sdks/packages/python/examples/01_list_devices.py
#!/usr/bin/env python3
"""01 — List attached devices.

The "hello world" of the SDK: construct an :class:`~go_ios_sdk.IosClient`, ask the
daemon which devices are attached, and print each one's udid (serial number) and
a couple of identifying properties.

``client.devices.list()`` returns the raw ``DeviceList`` envelope from the daemon
(a dict with a ``deviceList`` array). ``client.devices.udids()`` is a convenience
that pulls just the serial numbers out of that envelope.

Run it::

    export GO_IOS_API_KEY=your-key
    uv run python examples/01_list_devices.py
"""

from __future__ import annotations

from _common import base_url, print_header, require_api_key

from go_ios_sdk import IosClient


def main() -> None:
    print_header("01 list_devices")
    api_key = require_api_key()

    # The client is a context manager; leaving the ``with`` closes the underlying
    # httpx connection pool. base_url + api_key come from the environment.
    with IosClient(base_url=base_url(), api_key=api_key) as client:
        envelope = client.devices.list()
        device_list = envelope.get("deviceList", []) or []

        print(f"attached devices: {len(device_list)}")
        for entry in device_list:
            props = entry.get("properties", {}) or {}
            udid = props.get("serialNumber", "<unknown>")
            name = props.get("DeviceName", props.get("ProductType", ""))
            version = props.get("ProductVersion", "")
            print(f"  - {udid}  {name} {version}".rstrip())

        # udids() is the shortcut you'll reach for most often.
        print(f"udids(): {client.devices.udids()}")


if __name__ == "__main__":
    main()
